File: 1184-car-pooling/1184-car-pooling.cpp
class Solution {
public:
    bool carPooling(vector<vector<int>>& trips, int capacity) {
        // prefix sum se ho jaega lets try

        // sabse pehle maxm kaha tak trip hogi nikalo, location since utne size ka vector chahiye rahega
        int n=-1;
        for(auto it:trips){
            n=max({n,it[2],it[1]});
        }

        // n+2 size ka banao
        vector<int> prefix(n+2,0);
        // prefix[i]=4 means dist= i par car me 4 passenger baithenge best case me

        // iterate krke add krte jao
        for(auto it:trips){
            int start=it[1];
            int end=it[2];
            int ppl=it[0];

            // prefix me passengers add krdo
            prefix[start]+=ppl;
            prefix[end]-=ppl;
        }

        // ab prefix array bana do
        for(int i=1;i<n+2;i++){
            prefix[i]+=prefix[i-1];
        }

        // agar kahi par bi overall passenger cnt> capacity hua toh return false
        for(auto it:prefix){
            if(it>capacity) return false;
        }
        return true;
    }
};

File: 1184-car-pooling/1184-car-pooling.java
class Solution {
    public boolean carPooling(int[][] trips, int capacity) {
        int[] prefix=new int[1002];

        for(int[] currtrip:trips){
            int passenger=currtrip[0];
            int currstart=currtrip[1];
            int currend=currtrip[2];

            prefix[currstart]+=passenger;
            prefix[currend]-=passenger;
        }

        for(int i=1;i<1002;i++){
            prefix[i]+=prefix[i-1];
        }

        // prefix[i]=4 meaning-> i th pt par 4 passenger baithe hai
        for(int it:prefix){
            if(it>capacity) return false;
        }
        return true;

    }
}